controlplane/kubeadm/internal/controllers/status.go

```diff
@@ -43,8 +43,7 @@ func (r *KubeadmControlPlaneReconciler) updateStatus(ctx context.Context, contro
 
 	// set basic data that does not require interacting with the workload cluster
 	controlPlane.KCP.Status.Replicas = replicas
-	controlPlane.KCP.Status.ReadyReplicas = 0
-	controlPlane.KCP.Status.UnavailableReplicas = replicas
+	controlPlane.KCP.Status.UnavailableReplicas = replicas - controlPlane.KCP.Status.ReadyReplicas
```

Comment:
Note: This change avoids setting controlPlane.KCP.Status.ReadyReplicas from none 0 to 0 when GetWorkloadCluster(ctx) hits ErrClusterLocked or other error.

Reply:
BTW：https://github.com/kubernetes-sigs/cluster-api/blob/release-1.5/controlplane/kubeadm/internal/controllers/status.go#L93 uses replicas rather than desiredReplicas to calculate UnavailableReplicas. Not sure if it's a bug or on purpose. cc @fabriziopandini please help take a look.

Reply:
it would be odd to have two different ways to calculate the unavailable replicas within the same status update function.

Reply:
hi @neolit123, actually it uses the same way to calculate the unavailable replicas. The line 46 is only for covering the case that KCP.Status.ReadyReplicas is 0 since the default value of UnavailableReplicas is 0 when KCP is newly created. If ReadyReplicas is not 0, UnavailableReplicas should equal to `replicas - controlPlane.KCP.Status.ReadyReplicas`.
```
line 45-46:
   controlPlane.KCP.Status.Replicas = replicas
   controlPlane.KCP.Status.UnavailableReplicas = replicas - controlPlane.KCP.Status.ReadyReplicas

line 91-92:
	controlPlane.KCP.Status.ReadyReplicas = status.ReadyNodes
	controlPlane.KCP.Status.UnavailableReplicas = replicas - status.ReadyNodes
```

Reply:
hi @neolit123 @fabriziopandini could you please help take some time to move this forward? We need this patch to fix the issue hit in our project based on CAPI.

Reply:
might be worth adding some comments to avoid confusing.

Reply:
Sure. Added a comment to explain the logic. @neolit123 @fabriziopandini could you please take a look ?

Reply:
I'm not sure if the current logic is correct. Basically we take new information "len(controlPlane.Machines)" and then substract the current ReadyReplicas. Isn't UnavailableReplicas wrong if "len(controlPlane.Machines)" changed since we last wrote `UnavailableReplicas`?

Would it be better to more explicitly handle the two cases if our goal is not change UnavailableReplicas when we can't communicate with the workload cluster?

e.g.:
```go
	// Set basic data that does not require interacting with the workload cluster.
	if controlPlane.KCP.Status.Replicas == 0 { // .Status.Replicas is only ever 0 on the first reconcile.
		// Set UnavailableReplicas to `replicas` when KCP is newly created, otherwise keep it unchanged. So as to avoid updating it when unable to get the workload cluster.
		controlPlane.KCP.Status.UnavailableReplicas = replicas
	}
	controlPlane.KCP.Status.Replicas = replicas
```

Reply:
But anyway this change comes down to leaving `UnavailableReplicas` unchanged if we can't communicate with the workload cluster. I don't know if we want that. But probably it's better to leave it unchanged than guessing that all replicas are unavailable.

Reply:
Yup. `controlPlane.KCP.Status.UnavailableReplicas = replicas - controlPlane.KCP.Status.ReadyReplicas` already implies setting status.UnavailableReplicas to replicas if status.ReadyReplicas is 0.  In future we could consider adding more complex logic (e.g. >= failed times or minutes) when it can't communicate with the workload cluster.

Reply:
> already implies setting status.UnavailableReplicas to replicas if status.ReadyReplicas is 0

That's correct. But it will change the value of `UnavailableReplicas` if `replicas := int32(len(controlPlane.Machines))` changes since the last reconcile.

I would prefer to keep both UnavailableReplicas and ReadyReplicas absolutely unchanged in this case

Reply:
@sbueringer Got your point. If doing it this way, we may also need to move `controlPlane.KCP.Status.Replicas = replicas` to before line 96 `controlPlane.KCP.Status.ReadyReplicas = status.ReadyNodes`, otherwise only Status.Replicas increases or decreases by 1 and other Replicas fields unchanged (which is not reasonable). Even though the conditions are updated while no Replicas fields are updated (which is not reasonable either).

If updating both Status.Replicas and Status.UnavailableReplicas before getting ClusterStatus like this PR currently behaviors,  
1) it sounds reasonable when replicas changes from 3 to 4 (when rolling out KCP):  Replicas -> 3 & UnavailableReplicas -> 1
2) but a little unreasonable when replicas changes from 3 to 2 (when rolling out KCP):  Replicas -> 2 & UnavailableReplicas -> -1 (anyway we can change the PR to enforce UnavailableReplicas >= 0)

Reply:
Good point. Let's do the following:
1. in the create case: let's set .status.replicas, .status.unavailableReplicas, .status.readyReplicas (we'll set readyReplicas to 0 but let's do it anyway for clarity)
2. Then we set .stauts.version & conditions (l.55-82)
3. Then let's get the cluster status and only afterwards modify the status replica fields

I think it makes sense to update the conditions in 2 as they are all based on information from the control plane cluster.

Reply:
@sbueringer Got it.  I pushed a commit to use `Status.UnavailableReplicas = desiredReplicas- status.ReadyNodes` which should solve the problems we discussed above. Please take a look if it is. If not, I will make the changes as you described, but not sure why we need to  move down the update of status updatedReplicas because it doesn't depend on the workload cluste status: `controlPlane.KCP.Status.UpdatedReplicas = int32(len(controlPlane.UpToDateMachines()))`

Reply:
The idea would have been to not update any status replica fields if we can't communicate with the workload cluster to not get them into an inconsistent state with each other. I'll think a bit about it.

Reply:
Yup, I agree that the idea way would have been to not update any **status replica fields and conditions fields** if we can't communicate with the workload cluster.